<a href="https://colab.research.google.com/github/jugernaut/Prometeo/blob/master/04_M%C3%A9todosNum%C3%A9ricos/01_M%C3%A9todosIterativos/01_MetodosIterativos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Métodos iterativos
### Trabajo realizado con el apoyo del Programa UNAM-DGAPA-PAPIME PE101019

*   Autor: 
    - Ingrid Pamela Ruiz Puga

*   Rev: vier may 14 14:09:45 CDT 2021

## Introducción

Los métodos iterativos son aquellos que nos ayudan resolver diversos problemas matemáticos, por ejemplo podemos obtener una aproximación a la solución de un sistema de ecuaciones. Se puede decir que progresivamente va calculando aproximaciones a la solución de un problema. El procedimiento se repite sobre las nuevas aproximaciones para mejorar la solución aproximada hasta que el resultado más reciente satisfaga ciertos requisitos.

A diferencia de los métodos directos, en los cuales se debe terminar el proceso para tener la respuesta y suelen tener inconvenientes en aplicaciones de sistemas de grandes dimensiones, en los métodos iterativos se puede suspender el proceso al termino de una iteración.

Hay que destacar que este tipo de métodos crea una sucesión de puntos que van a converger a la sucesión de puntos solución.


Podemos generalizar un método iterativo en los siguientes pasos:

* Inicia con una solución aproximada (semilla).
* Ejecuta una serie de cálculos para obtener o construir una mejor aproximación partiendo de la semilla. 
* Se repite el paso anterior pero usando como semilla la aproximación obtenida.

Consideraremos los sistemas lineales ya conocidos:

Sea un sistema de ecuaciones lineales

\begin{align}
a_{11}x_1 + ... + a_{1n}x_n &= b_1\\
a_{21}x_2 + ... + a_{2n}x_n &= b_2\\
a_{31}x_3 + ... + a_{3n}x_n &= b_3\\
...\\
a_{n1}x_1 + ... + a_{nn}x_n &= b_n
\end{align}

Este tipo de sistemas de ecuaciones aparecen en muchas aplicaciones cotidianas como procesamiento de señales, simulación, analísis y procesamiento de datos espaciales.



## Método de Jacobi

Enlistaremos los pasos a seguir para resolver un sistema de ecuaciones por el método de Jacobi.

* Despejar de cada ecuación una incognita.
* Sustituir la semilla en las ecuaciones para obtener la aproximación a cada punto.
* La aproximación obtenida será la nueva semilla.
* Repetir el proceso hasta el punto de paro.

De manera matricial vemos el método de Jacobi como:

$$M=D$$ 
$$N=-(L+U)$$

Donde $D$ es la matriz diagonal de la matriz asociada al sistema de ecuaciones.

$L$ es la matriz tal que $l_{ij}=a_{ij}$ si $i<j$ y $l_{ij}=0$ en caso contrario.

$U$ es la matriz tal que $u_{ij}=a_{ij}$ si $i>j$ y $u_{ij}=0$ en caso contrario.

Con lo cual 
$$A=D+L+U = M-N$$ y el sistema 
$$Mx =Nx +b$$
es equivalente a resolver 
$$Dx = -(L+U)x +b$$
Así, la sucesión se construye partiendo de un valor inicial $x_0$ y definiendo
$$ Dx_{k+1} = -(L+U)x_k +b$$

para $k\geq0$






### Ejemplo 1
Consideremos el siguiente sistema de ecuaciones:

\begin{align}
2x -y &= 9\\
x +6y -2z &= 15\\
4x -3y +8z &= 1\\
\end{align}

Despejamos las incognitas de las ecuaciones:
\begin{align}
x  &= (9 +y_0 )/2\\
y  &= (15 -x_0 +2z_0)/6\\
z &= (1-4x_0+3y_0)/8\\
\end{align}


Teniendo el punto inicial $v_0 = (x_0, y_0, z_0)$.
Suponiendo que el punto inicial sea $v_0 = (0,0,0)$ sustituimos y calculamos el siguiente punto.

\begin{align}
x  &= (9 +0 )/2\\
y  &= (15 -0 +2(0))/6\\
z &= (1-4(0)+3(0))/8\\
\end{align}

Obtenemos
$$v_1 = (4.5,2.5,0.125)$$

Continuamos a la siguiente iteración usando $v_1$ como semilla.

\begin{align}
x  &= (9 + (2.5)/2\\
y  &= (15 -4.5 +2(0.125))/6\\
z &= (1-4(4.5)+3(2.5))/8\\
\end{align}

Y obtenemos el punto:
$$v_2 = (5.75,1.7916,-1.1875)$$

Se realiza el procedimiento recursivamente hasta parar en alguna iteración.

En este caso las aproximaciones convergen en el punto $(5,1,-2)$



### Programación del método

In [ ]:
#Importamos libreria
import numpy as np

In [ ]:
#Función despeja la matriz asociada al sistema de ecuaciones
def coeficientes(A,b):
    S = np.zeros([A.shape[0], A.shape[1]+1])
    W = np.copy(A)-np.diag(np.diag(A))
    for i in range(A.shape[0]):
        S[i][0]=b[i]/np.diag(A)[i]
    for i in range(S.shape[0]):
        for j in range(1,S.shape[1]):
            S[i][j]=-W[i][j-1]/np.diag(A)[i]
    return S

#Función para evaluar las ecuaciones despejadas
def ecuacion(C,var,r):
    v=0
    for i in range(1,C.shape[1]):
        v+=(C[r][i]*var[i-1])
    v=v+C[r][0]
    return v

#Función que realiza el método de jacobi
def jacob(C, var, iteraciones):
    global p
    if p < iteraciones:
        val=[]
        for i in  range(C.shape[0]):
            t=ecuacion(C,var,i)
            val.append(t)
            #print(var)
        p+=1
        print("Iteración",p,"x:",val[0]," ", "y:",val[1]," ","z:",val[2])
        jacob(C, val, iteraciones)


In [ ]:
#Definimos la matriz del sistema de ecuaciones
A= np.array([[2.0, -1.0, 0.0],[1.0, 6.0, -2.0],[4.0, -3.0, 8.0]])

#Valor solución a cada ecuación
b=np.array([9.0,15.0,1.0])

#Punto inicial (semilla)
var=np.array([0.0, 0.0 ,0.0 ])

#Despejamos ecuación
C=coeficientes(A,b)

#Variable conteo inicial de iteraciones
p=0

#Número de iteraciones (criterio de paro)
iteraciones=10
jacob(C, var , iteraciones)

Iteración 1 x: 4.5   y: 2.5   z: 0.125
Iteración 2 x: 5.75   y: 1.7916666666666665   z: -1.1875
Iteración 3 x: 5.395833333333333   y: 1.1458333333333335   z: -2.078125
Iteración 4 x: 5.072916666666667   y: 0.9079861111111112   z: -2.1432291666666665
Iteración 5 x: 4.953993055555555   y: 0.9401041666666667   z: -2.070963541666667
Iteración 6 x: 4.970052083333333   y: 0.9840133101851851   z: -1.9994574652777777
Iteración 7 x: 4.992006655092593   y: 1.005172164351852   z: -1.9910210503472223
Iteración 8 x: 5.002586082175926   y: 1.0043252073688271   z: -1.9940637659143516
Iteración 9 x: 5.002162603684414   y: 1.0015477309992287   z: -1.9996710883246527
Iteración 10 x: 5.000773865499614   y: 0.9997492032777135   z: -2.000500902717496


## Método Gauss-Seidel

Enlistaremos los pasos a seguir para resolver un sistema de ecuaciones por el método de Gauss-Seidel .

* Despejar de cada ecuación una incognita.
* Sustituir la semilla en la primera ecuación y se actualiza el valor obtenido en la semilla para sustituir en la siguiente ecuación y asi recursivamente.
* La aproximación obtenida será la nueva semilla para la siguiente iteración.
* Repetir el proceso hasta el punto de paro.

De manera matricial vemos el método de Jacobi como:

$$M=D+L$$ 
$$N=-U$$

Donde $D$ es la matriz diagonal de la matriz asociada al sistema de ecuaciones.

$L$ es la matriz tal que $l_{ij}=a_{ij}$ si $i<j$ y $l_{ij}=0$ en caso contrario.

$U$ es la matriz tal que $u_{ij}=a_{ij}$ si $i>j$ y $u_{ij}=0$ en caso contrario.

Con lo cual 
$$Mx =Nx +b$$
es equivalente a resolver 
$$(D+L)x = -Ux +b$$
Así, la sucesión se construye partiendo de un valor inicial $x_0$ y definiendo
$$ (D+L)x_{k+1} = -Ux_k +b$$

para $k\geq0$


### Ejemplo 2
Consideremos el mismo sistema de ecuaciones:

\begin{align}
2x -y &= 9\\
x +6y -2z &= 15\\
4x -3y +8z &= 1\\
\end{align}

Despejamos las incognitas de las ecuaciones:
\begin{align}
x  &= (9 +y_0 )/2\\
y  &= (15 -x_0 +2z_0)/6\\
z &= (1-4x_0+3y_0)/8\\
\end{align}


Teniendo el punto inicial $v_0 = (x_0, y_0, z_0)$.
Suponiendo que el punto inicial sea $v_0 = (0,0,0)$ sustituimos y calculamos el siguiente punto:

\begin{align}
x  &= (9 +0 )/2 = 4.5\\
y  &= (15 -4.5 +2(0))/6 = 1.75\\
z &= (1-4(4.5)+3(1.75))/8 = -1.46\\
\end{align}

Obtenemos
$$v_1 = (4.5,1.75,-1.46)$$

Continuamos a la siguiente iteración usando $v_1$ como semilla.

\begin{align}
x  &= (9 + (4.5)/2 = 5.375\\
y  &= (15 -5.375 +2(-1.46))/6 = 1.11\\
z &= (1-4(5.375)+3(1.11))/8 = -2.14\\
\end{align}

Y obtenermos el punto:
$$v_2 = (5.375,1.11,-2.14)$$

Se realiza el procedimiento recursivamente hasta parar en alguna iteración.

En este caso las aproximaciones convergen en el punto $(5,1,-2)$ más rápido que con el método de Jacobi.


### Programación del método

In [ ]:
#Función despeja la matriz asociada al sistema de ecuaciones
def coeficientes(A,b):
    S = np.zeros([A.shape[0], A.shape[1]+1])
    W = np.copy(A)-np.diag(np.diag(A))
    for i in range(A.shape[0]):
        S[i][0]=b[i]/np.diag(A)[i]
    for i in range(S.shape[0]):
        for j in range(1,S.shape[1]):
            S[i][j]=-W[i][j-1]/np.diag(A)[i]
    return S

#Función para evaluar las ecuaciones despejadas
def ecuacion(C,var,r):
    v=0
    for i in range(1,C.shape[1]):
        v+=(C[r][i]*var[i-1])
    v=v+C[r][0]
    return v

#Función que realiza el método de Gauss-Seidel
def GaussSeidel(C, var, iteraciones):
    global p
    if p < iteraciones:
        val=[]
        for i in  range(C.shape[0]):
            t=ecuacion(C,var,i)
            val.append(t)
            var[i]=t
            #print(var)
        p+=1
        print("Iteracion {}:".format(p), val)
        GaussSeidel(C, val, iteraciones)


In [ ]:
#Definimos la matriz del sistema de ecuaciones
A= np.matrix([[10.0,0.0,-1.0],[4.0,12.0,-4.0],[4.0,4.0,10.0]])

#Valor solución a cada ecuación
b=np.array([-1.0,8.0,4.0])

#Punto inicial (semilla)
var=np.array([1.0, 2.0 ,0.0 ])

#Despejamos ecuación
C=coeficientes(A,b)

#Variable conteo inicial de iteraciones
p=0

#Número de iteraciones (criterio de paro)
iteraciones=10
GaussSeidel(C, var , iteraciones)

Iteracion 1: [-0.1, 0.7, 0.16000000000000006]
Iteracion 2: [-0.08399999999999999, 0.748, 0.13440000000000002]
Iteracion 3: [-0.08656, 0.74032, 0.138496]
Iteracion 4: [-0.0861504, 0.7415488, 0.13784064000000001]
Iteracion 5: [-0.086215936, 0.7413521919999999, 0.13794549760000008]
Iteracion 6: [-0.08620545024, 0.74138364928, 0.137928720384]
Iteracion 7: [-0.0862071279616, 0.7413786161152, 0.13793140473856003]
Iteracion 8: [-0.086206859526144, 0.741379421421568, 0.1379309752418304]
Iteracion 9: [-0.08620690247581697, 0.7413792925725491, 0.13793104396130712]
Iteracion 10: [-0.0862068956038693, 0.7413793131883921, 0.13793103296619091]


### Ejercicio
Utiliza el método de Gauss-Seidel en *Python* para solucionar el sistema de ecuaciones del ejemplo 1 y compara qué método converge más rápido.

## Referencias
Para profundizar en los temas tratados en el presente documento se pueden consultar las siguientes fuentes:

* https://www.mty.itesm.mx/dmti/materias/ma2008/lecturas/ma2008-09a.pdf
* http://matematicas.unex.es/~trinidad/MM/tema2.pdf